Rock paper scissors game

sopa40 opened the pull request:
Rock paper scissors game from scratch

marynasergiyenko committed:
Update README.md

ekovalyov committed:
Create README.md

sopa40 committed:
Add game skeleton

Asking user if he wants to play

Signed-off-by: Nazar Sopiha <nazar.sopiha@gmail.com>

sopa40 committed:
Add game logic

Game logic according to wiki rules using switch-case methods. TODO in comments

Signed-off-by: Nazar Sopiha <nazar.sopiha@gmail.com>

sopa40 committed:
Change code style

Signed-off-by: Nazar Sopiha <nazar.sopiha@gmail.com>

sopa40 committed:
New file structure

Signed-off-by: Nazar Sopiha <nazar.sopiha@gmail.com>

ReAlUA commented on scissors/main.c:
```diff
@@ -0,0 +1,94 @@
+    char usr_key = 'y';
+    printf("Do you want to play the game? y or n \n");
+    scanf(" %c", &usr_key);
+    switch (usr_key) {
```
As for me, any answer except `tolower(usr_key) == 'y'` is "I won't play".
ESC or 'q' is not an *error* here.

ReAlUA commented on scissors/main.c:
```diff
@@ -0,0 +1,94 @@
+    fflush(stdin);
+    /* 0 == rock, 1 == paper, 2 == scissors */
+    int mchn_optn = rand()%3;
+    char usr_optn;
```
Use [Linux kernel coding style](https://www.kernel.org/doc/html/latest/process/coding-style.html) (variable declaration placement, indentation, ..,)

ReAlUA replied:
You can use `linux-stable/scripts/checkpatch.pl` in file mode to detect all the issues.

ReAlUA commented on scissors/main.c:
```diff
@@ -0,0 +1,94 @@
+    printf("You chose %c, I chose %d\n", usr_optn, mchn_optn);
+    switch (mchn_optn) {
+        case 0:
+            if (usr_optn == 'r')
```
This long structure with many indentation levels is not so good for reading and maintenance.
Also, there is no connection between number-coded and char-coded options.
If you convert user input to number or machine choice to character, you can use just `mchn_optn == usr_optn` to detect a nobody wins case and use only one comparison in `case`s.

sopa40 committed:
Fix style to match kernel

Signed-off-by: Nazar Sopiha <nazar.sopiha@gmail.com>

ReAlUA commented on scissors/main.c:
```diff
@@ -0,0 +1,104 @@
+		return 0;
+	}
+
+	return -1;
```
Unreachable code due to the `default` branch above

ReAlUA commented on scissors/main.c:
```diff
@@ -0,0 +1,104 @@
+		else
+			printf("You won!\n");
+	} else if (dcsn == -1) {
+		printf("Unknown option. Error\n");
```
Unreachable code, `invite_to_play` never returns -1

ReAlUA commented on scissors/main.c:
```diff
@@ -0,0 +1,104 @@
+int play(void)
+{
+	fflush(stdin);
+	/* 0 == rock, 1 == paper, 2 == scissors */
```
Add a enum. It's too hard to analyze code in switch below.

sopa40 committed:
Change game logic

Add two enums: result and option to make code more,
fix unreachable parts of code,
Change input convertion because atoi() needs a string

Signed-off-by: Nazar Sopiha <nazar.sopiha@gmail.com>

ReAlUA commented on scissors/main.c:
```diff
@@ -0,0 +1,104 @@
+		return -1;
+	}
+
+	usr_optn = atoi(&usr_optn_input);
```
Sorry, I didn't notice this at first.
`usr_optn_input` is a single character, not string. It works just thanks to luck.
Either convert `usr_optn_input` into two-character array with explicit 0 in the second element or use another way to convert into a number, say, `usr_optn = usr_optn_input - '0'`
